### ISSUE TITLE
How do you place objects on a plane randomly?

### ISSUE THREAD

**Issue by RealRohan** [google-ar/sceneform-android-sdk#490]
I want to build an AR game using Sceneform where an object is placed on the plane on a timer (ex: every 1 second, an object is placed). I also want it to be placed at a random position on the detected plane. I only know how to place an object using a HitTest, but that places it only where and when the user taps. I want the placement to be controlled by the app as this is a game.
Thanks!

**Comment by claywilkinson** [comment]
You can create anchors a any point, you could use the *y* position of the plane as the fixed y value and then vary  *x* and *z* within the plane polygon.

**Comment by RealRohan** [comment]
> You can create anchors a any point, you could use the _y_ position of the plane as the fixed y value and then vary _x_ and _z_ within the plane polygon.

@claywilkinson an example of this would be very useful as I am a new Sceneform developer. Thanks!

**Comment by claywilkinson** [comment]
Based on the [HelloSceneform sample](https://github.com/google-ar/sceneform-android-sdk/tree/master/samples/hellosceneform),

**Note:** You'll need to add some member variables - but I think you'll figure it out :)
1. Add an update listener that is called every frame in `onCreate()`
      ```
       arFragment.getArSceneView().getScene().addOnUpdateListener(this::onFrame);
 2.  Add the `onFrame()` method
```
private void onFrame(FrameTime frameTime) {
    // Keep track of the first valid plane detected, update it 
    // if the plane is lost or subsumed.

    if (firstPlane == null || firstPlane.getTrackingState() != TrackingState.TRACKING) {
      Collection<Plane> planes = arFragment.getArSceneView().getSession()
              .getAllTrackables(Plane.class);
      if (!planes.isEmpty()) {
        firstPlane = planes.iterator().next();
      }
      return;
    }
    if (firstPlane.getSubsumedBy() != null) {
      firstPlane = firstPlane.getSubsumedBy();
    }

    // Every 2 seconds move the node to a different spot.
    if (nodeAge > 2) {
      randomPlacedCube(firstPlane);
      nodeAge = 0;
    }
    nodeAge += frameTime.getDeltaSeconds();
  }

```

3. Add the `randomPlacedCube()` method.
```
private static Random rand = new Random();

  private void randomPlacedCube(Plane plane) {
    //find a random spot on the plane in the X
    // The width of the plan is 2*extentX in the range center.x +/- extentX
    float maxX = plane.getExtentX() * 2;
    float randomX = (maxX * rand.nextFloat()) - plane.getExtentX();

    float maxZ = plane.getExtentZ() * 2;
    float randomZ = (maxZ * rand.nextFloat()) - plane.getExtentZ();

    Pose pose = plane.getCenterPose();
    float[] translation = pose.getTranslation();
    float[] rotation = pose.getRotationQuaternion();

    translation[0] += randomX;
    translation[2] += randomZ;
    pose = new Pose(translation, rotation);

    Anchor anchor = arFragment.getArSceneView().getSession().createAnchor(pose);

    if (anchorNode == null) {
      cubeRenderable =
              ShapeFactory.makeCube(Vector3.one().scaled(.25f),
                      new Vector3(.125f, 0, .125f), andyRenderable.getMaterial());
      node = new Node();
      node.setRenderable(cubeRenderable);
      anchorNode = new AnchorNode(anchor);
      anchorNode.addChild(node);
      anchorNode.setParent(arFragment.getArSceneView().getScene());
    } else {
      anchorNode.getAnchor().detach();
      anchorNode.setAnchor(anchor);
    }
  }
```

**Comment by RealRohan** [comment]
@claywilkinson thank you so much!

**Comment by RealRohan** [comment]
@claywilkinson I need a bit of help again... I adapted your code and put it into the HelloSceneform sample project, but it doesn't recognize "NodeAge" and "anchorNode." I can't figure these 2 out :(. Here's my MainActivity.java file:

**Note: instead of placing a cube, I have decided to just place the Andy/Android object in a random position every 2 seconds.**

package com.google.ar.sceneform.samples.hellosceneform;

import android.app.Activity;
import android.app.ActivityManager;
import android.content.Context;
import android.os.Build;
import android.os.Build.VERSION_CODES;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Gravity;
import android.view.MotionEvent;
import android.widget.Toast;
import com.google.ar.core.Anchor;
import com.google.ar.core.HitResult;
import com.google.ar.core.Plane;
import com.google.ar.core.Pose;
import com.google.ar.core.TrackingState;
import com.google.ar.sceneform.AnchorNode;
import com.google.ar.sceneform.FrameTime;
import com.google.ar.sceneform.Node;
import com.google.ar.sceneform.rendering.ModelRenderable;
import com.google.ar.sceneform.ux.ArFragment;
import com.google.ar.sceneform.ux.TransformableNode;

import java.util.Collection;
import java.util.Random;


public class HelloSceneformActivity extends AppCompatActivity {
  private static final String TAG = HelloSceneformActivity.class.getSimpleName();
  private static final double MIN_OPENGL_VERSION = 3.0;

  private ArFragment arFragment;
  private ModelRenderable andyRenderable;
  private Plane plane;

  @Override
  @SuppressWarnings({"AndroidApiChecker", "FutureReturnValueIgnored"})
  // CompletableFuture requires api level 24
  // FutureReturnValueIgnored is not valid
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    if (!checkIsSupportedDeviceOrFinish(this)) {
      return;
    }

    setContentView(R.layout.activity_ux);
    arFragment = (ArFragment) getSupportFragmentManager().findFragmentById(R.id.ux_fragment);

    arFragment.getArSceneView().getScene().addOnUpdateListener(this::onFrame);

    // When you build a Renderable, Sceneform loads its resources in the background while returning
    // a CompletableFuture. Call thenAccept(), handle(), or check isDone() before calling get().
    ModelRenderable.builder()
        .setSource(this, R.raw.andy)
        .build()
        .thenAccept(renderable -> andyRenderable = renderable)
        .exceptionally(
            throwable -> {
              Toast toast =
                  Toast.makeText(this, "Unable to load renderable", Toast.LENGTH_LONG);
              toast.setGravity(Gravity.CENTER, 0, 0);
              toast.show();
              return null;
            });
  }

  private void onFrame(FrameTime frameTime) {
      if (nodeAge > 1) {
          placeAndy(plane);
          nodeAge = 0;
      }
      nodeAge += frameTime.getDeltaSeconds();
  }

  private static Random rand = new Random();

  private void placeAndy(Plane plane) {
      float maxX = plane.getExtentX() * 2;
      float randomX = (maxX * rand.nextFloat()) - plane.getExtentX();

      float maxZ = plane.getExtentZ() * 2;
      float randomZ = (maxZ * rand.nextFloat()) - plane.getExtentZ();

      Pose pose = plane.getCenterPose();
      float[] translation = pose.getTranslation();
      float[] rotation = pose.getRotationQuaternion();

      translation[0] += randomX;
      translation[2] += randomZ;
      pose = new Pose(translation, rotation);

      Anchor anchor = arFragment.getArSceneView().getSession().createAnchor(pose);

      if (anchorNode == null) {
          anchorNode.setParent(arFragment.getArSceneView().getScene());

          // Create the andy and add it to the anchor.
          Node andy = new TransformableNode(arFragment.getTransformationSystem());
          andy.setParent(anchorNode);
          andy.setRenderable(andyRenderable);
      }
  }

  public static boolean checkIsSupportedDeviceOrFinish(final Activity activity) {
    if (Build.VERSION.SDK_INT < VERSION_CODES.N) {
      Log.e(TAG, "Sceneform requires Android N or later");
      Toast.makeText(activity, "Sceneform requires Android N or later", Toast.LENGTH_LONG).show();
      activity.finish();
      return false;
    }
    String openGlVersionString =
        ((ActivityManager) activity.getSystemService(Context.ACTIVITY_SERVICE))
            .getDeviceConfigurationInfo()
            .getGlEsVersion();
    if (Double.parseDouble(openGlVersionString) < MIN_OPENGL_VERSION) {
      Log.e(TAG, "Sceneform requires OpenGL ES 3.0 later");
      Toast.makeText(activity, "Sceneform requires OpenGL ES 3.0 or later", Toast.LENGTH_LONG)
          .show();
      activity.finish();
      return false;
    }
    return true;
  }
}

**Comment by claywilkinson** [comment]
nodeAge is the age of the current position, anchorNode is also a member.
Add a member fields
private float nodeAge;
private AnchorNode anchorNode;
